scheduler/src/cook/mesos/scheduler.clj

```diff
@@ -407,104 +416,239 @@
-                           [[:job/allowed-to-start? (:job-id task-info)]
-                            {:db/id (d/tempid :db.part/user)
-                             :job/_instance (:job-id task-info)
-                             :instance/task-id (:task-id task-info)
-                             :instance/hostname (:hostname offer)
-                             :instance/start-time (now)
-                             ;; NB command executor uses the task-id
-                             ;; as the executor-id
-                             :instance/executor-id (get-in
-                                                     task-info
-                                                     [:executor :executor-id]
-                                                     (:task-id task-info))
-                             :instance/slave-id (:slave-id offer)
-                             :instance/progress 0
-                             :instance/status :instance.status/unknown
-                             :instance/preempted? false}])
-                         task-infos))
-              (log/info "Matched offer" offer "to tasks" task-infos)
+                 (vec (apply concat upgrade-txns task-txns)))
+              (log/info "Launching" (count task-txns) "tasks")
+              (log/info "Upgrading" (count (:upgrade-backfill processed-matches)) "tasks from backfilled to proper")
+              (log/debug "Matched tasks" task-txns)
               ;; This launch-tasks MUST happen after the above transaction in
               ;; order to allow a transaction failure (due to failed preconditions)
               ;; to block the launch
-              (meters/mark! scheduler-offer-matched (count (:ids offer)))
-              (meters/mark! matched-tasks (count task-infos))
+              (meters/mark! scheduler-offer-matched
+                            (->> matches
+                                 (mapcat (comp :id :offer :leases))
+                                 (distinct)
+                                 (count)))
+              (meters/mark! matched-tasks (count task-txns))
               (meters/mark! matched-tasks-cpus (:cpus match-resource-requirements))
               (meters/mark! matched-tasks-mem (:mem match-resource-requirements))
-              (mesos/launch-tasks
-                driver
-                (:ids offer)
-                (mapv #(dissoc % :job-id :num-ports) task-infos))))))
+              (doseq [{:keys [tasks leases]} matches
+                      :let [offers (mapv :offer leases)
+                            task-infos (mapv (fn [^TaskAssignmentResult task]
+                                               (:task-info (.getRequest task)))
+                                             tasks)
+                            slave-id (:slave-id (first offers))]]
+                ;TODO if the ports are in the offer, then Fenzo is identifying the ports with a list .getAssignedPorts on the TaskAssignmentResult
```

Comment:
We're still not handling ports at the very end, but the entire complex bits of the code path are handled. We should finish this before merge.